src/overlay/OverlayManagerImpl.cpp

```diff
@@ -224,8 +243,9 @@ OverlayManagerImpl::connectToMorePeers(int max)
         {
             continue;
         }
-        if (getAuthenticatedPeersCount() >=
-            mApp.getConfig().TARGET_PEER_CONNECTIONS)
+        // we always try to connect to preferred peers
```

Comment:
what will happen here if we have all available connections occupied with preferred peers and pr is a non-connected preferred?

Reply:
in this case it will connect anyways and will take the number of connections above TARGET at this stage (as a pending connection).
The logic in `OverlayManagerImpl::acceptAuthenticatedPeer` will deal with the connection properly later (kicking out other nodes if it can - enforcing `MAX_PEER_CONNECTIONS`).
If `MAX_PEER_CONNECTIONS` is mis-configured (like cannot accommodate for all preferred peers, including by host and by key), it will try every few seconds to reconnect, but that seems fine.